(SLV-422) Add PR testing

johnduarte opened the pull request:
This PR incorporates includes the addition of a `.travis.yml` file to support PR testing along with the additional structure to provide testing for:

* rubocop
* puppet-lint
* rspec

johnduarte committed:
(SLV-422) Downgrade stdlib to 4.25.1

This downgrade resolves dependency issues while maintaining
compatibility with puppet > 7.

johnduarte committed:
(MAINT) Remove commented module requirement

johnduarte committed:
(SLV-422) Update modules to resolve dependencies

This commit updates module version definitions in the Puppetfile to
resolve all dependency issues.

johnduarte committed:
(SLV-422) Add Gemfile for testing

johnduarte committed:
(SLV-422) Add rake task for puppet-lint, fix errs

This commit adds a Rakefile to provide the puppet-lint task.  Issues
discovered by running puppet-lint are also corrected.

johnduarte committed:
(SLV-422) Add rake task for rubocop, fix errs

This commit adds rubocop to the Rakefile along with a rubocop
configuration file.  Errors detected by rubocop are also corrected.

johnduarte committed:
(SLV-422) Remove metadata-json-lint from Gemfile

The metadata-json-lint was erroneously added to the Gemfile.  This
commit removes it.

johnduarte committed:
(SLV-422) Add spec test for module list

This commit adds a spec test that ensures that the installed module tree
has no invalid entries.  Rake tasks are added to install the modules and
to run the spec tests.

A .gitignore file is added to prevent any of the artifacts from this
process being added to version control.

johnduarte committed:
(SLV-422) Add .travis.yml for PR validation

This commit adds a configuration file for travis-ci to validate PRs
prior to merge.

The checks cover the following independently as part of a matrix:

* rubocop
* puppet-lint
* rspec tests

johnduarte committed:
(SLV-422) Add puppet-lint config

This commit adds a configuration file for puppet-lint.  Its primary
purpose is to ignore anything in the `modules` directory that is pulled
down for testing.

Most of the file consists of skipping validations that currently
generate errors.  These `--no-*` lines should be removed when the code
has been corrected to enable these checks to pass.

Ref: [SLV-674](https://tickets.puppetlabs.com/browse/SLV-674)

johnduarte committed:
(SLV-422) Move puppet-lint no checks into Rakefile

This commit deletes the .puppet-lint.rc file and moves all of the
configuration items from it into the Rakefile configuration for
puppet-lint.

Prior to this commit travis-ci did not pick up the configuration
settings.

billclaytor commented on .rubocop.yml:
```diff
@@ -0,0 +1,27 @@
+    - 'vendor/**/*'
+Layout/AlignHash:
+  EnforcedHashRocketStyle: table
+Style/StringLiterals:
```
GPLT uses the following based on our discussion of single vs. double quotes:
```
Style/StringLiterals:
  EnforcedStyle: double_quotes
```

Should we enforce double quotes here as well for consistency?

RandellP replied:
+1   yes please.

johnduarte replied:
The enforcement of StringLiterals and FrozenStringLiteralComment were disabled to prevent a massive change set as part of this PR.  This work has been ticketed in https://tickets.puppetlabs.com/browse/SLV-678 to update these configuration settings and fix the code accordingly.

billclaytor commented on Rakefile:
```diff
@@ -0,0 +1,34 @@
+                             arrow_on_right_operand_line
+                             case_without_default]
+end
+
```
GPLT includes a `pr` namespace with a `pr:check` task to allow running all of the pr checks with a single command. Should we do the same here?

RandellP replied:
I'm a fan of this as well.  Last time we debated it we said we would go that pr:check route unless we had stuff that got significant time gains from being parallel.  So unless that is a factor I vote for the single task that does whatever should be done.

johnduarte replied:
:-1: I am not a fan of this.  I tried it with all of the checks in one travis run and it made it difficult to resolve issues in a timely manner.  Since the travis run will exit on the first failure, if the rubocop fails, then any additional failures in the puppet-lint and/or spec tests are not discovered until the full assembly of tests is made to pass in turn.  In this case, it tripled the amount of travis runs needed in order to discover the issues that were part of this PR.

billclaytor replied:
I don't mind if Travis runs them separately but it is helpful when testing locally to have a single task to run rather than having to consult the travis file to see what will be run and then running those separately myself.

RandellP replied:
ok yeah, if it stops I guess we shouldn't do that in travis.  But having a single rake task for dev testing is still on my "would like to have" list. Wish we could somehow have both...

johnduarte replied:
@RandellP I have added a `test:all` task to do this.

billclaytor commented on .travis.yml:
```diff
@@ -0,0 +1,20 @@
+sudo: false
+bundler_args: --jobs 4 --retry 2 --without packaging documentation
+before_install:
+  - rm -f Gemfile.lock
```
Why are we including `Gemfile.lock` but removing it at this step? Could this result in things passing locally but failing when a PR is created?

johnduarte replied:
The `Gemfile.lock` is not included in this PR and is not currently included in the git repo.

billclaytor replied:
Sorry, of course, my mistake. I will blame it on a lack of coffee ;)

billclaytor commented on Puppetfile:
```diff
@@ -9,7 +9,7 @@ mod 'puppetlabs/yumrepo_core', '1.0.0'
 # Modules required to get a tomcat server up and running
 mod 'puppetlabs/tomcat', '2.5.0'
-mod 'puppetlabs/stdlib', '5.2.0'
+mod 'puppetlabs/stdlib', '4.25.1'
```
Why is this version being changed? Should there be a comment with an explanation for this version being specified?

johnduarte replied:
I have added a comment for the restriction @billclaytor

RandellP commented on .travis.yml:
```diff
@@ -0,0 +1,20 @@
+  - rm -f Gemfile.lock
+  - gem update --system && gem install bundler --no-document
+script:
+  - "bundle exec rake $CHECK"
```
per comment below on rake tasks.  Consider using the pr::check task so there is a unified way to run all the checks...

RandellP commented on manifests/profile/loadbalancer_exports.pp:
```diff
@@ -1,3 +1,16 @@
+#   haproxy balancemembers.
+#
+# @param haproxy_options
+#   Options to use fro the haproxy balancermember
```
typo: fro

RandellP commented on manifests/role/loadbalancer_exports.pp:
```diff
@@ -1,3 +1,16 @@
+#   loadbalancer_exports profile.
+#
+# @param haproxy_options
+#   Options to use fro the haproxy balancermember
```
typo: fro

johnduarte committed:
(SLV-422) Comment rubocop configs to be fixed

This commit adds comments to rubocop configuration settings that need to
be corrected to bring it into parity with gatling-puppet-load-test as
part of the work for
[SLV-678](https://tickets.puppetlabs.com/browse/SLV-678).

johnduarte committed:
(SLV-422) Fix typo in manifest docs

johnduarte committed:
(SLV-422) Add comment for stdlib version change

johnduarte committed:
(SLV-422) Add Rake task to run all tests

billclaytor reviewed (approved):
Looks good!
